### PR TITLE
expose document change events for package extensions

### PR DESCRIPTION
**Pull request by kevinushey**
### Intent

Addresses https://github.com/rstudio/rstudio/issues/13278.

### Approach

This PR adds two pieces:

1. An API for setting ghost text in a document;
2. A mechanism to allow R code to listen for document changes.

The second mechanism feels a bit dangerous, but this doesn't add anything that packages couldn't already do since package authors can already see files on the user's filesystem, and also view and modify text buffers in RStudio via functions in `rstudioapi`.

### Automated Tests

N/A

### QA Notes

N/A; to be verified by community.

### Documentation

N/A

### Checklist

- [x] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [x] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [x] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [x] This PR passes all local unit tests

### CODE REVIEW

File: src/cpp/session/CMakeLists.txt [hunk 1]
```diff
@@ -104,7 +104,7 @@ if(WIN32)
 endif()
 
 # include files
-file(GLOB_RECURSE SESSION_HEADER_FILES "*.h*")
+file(GLOB_RECURSE SESSION_HEADER_FILES CONFIGURE_DEPENDS "*.h*")
```
Comment (jgutman) [inline src/cpp/session/CMakeLists.txt:107]: What's this for?

Comment (kevinushey) [reply src/cpp/session/CMakeLists.txt:107]: From https://cmake.org/cmake/help/latest/command/file.html#filesystem:

> New in version 3.12: If the CONFIGURE_DEPENDS flag is specified, CMake will add logic to the main build system check target to rerun the flagged GLOB commands at build time. If any of the outputs change, CMake will regenerate the build system.

In other words, it lets you add new header files, and CMake will automatically detect those new files on build without needing to re-run the CMake generator. Essentially, a small quality-of-life developer change.

**Comment by jgutman** [comment]
Did you want to target Desert Sunflower or Ocean Storm?

File: src/gwt/src/org/rstudio/studio/client/workbench/views/console/shell/assist/RCompletionManager.java [hunk 2]
```diff
@@ -412,21 +411,44 @@ public boolean previewKeyDown(NativeEvent event)
 
       int keycode = event.getKeyCode();
       int modifier = KeyboardShortcut.getModifierValue(event);
+ 
+      // Handle ghost text. Ghost text handlers get higher priority over other
+      // completion handlers.
+      if (docDisplay_.hasGhostText())
+      {
+         // Let tab insert ghost text if configured to do so
+         if (keycode == KeyCodes.KEY_TAB && modifier == 0)
```
[the remark this answers is not in the document]
Comment (kevinushey) [reply src/gwt/src/org/rstudio/studio/client/workbench/views/console/shell/assist/RCompletionManager.java:420]: You're right, thanks for catching that!

**Pull request merged**